### ISSUE TITLE
Support runtime tables

### ISSUE THREAD

**Issue by querolita** [o1-labs/o1js#1859]
This is a feature that enables the user to define tables of data at runtime. In order for them to work, the desired tables must be configured at constraint system creation phase. Then, they are populated with values. And finally, the prover can make assertions using data from the table using lookup rows.

This feature has been enabled in Kimchi and Pickles. Some parts are missing still in order to be able to port them into o1js. This issue will illustrate the relevant functions, types, and bindings involved in all the 3 repos.

**Comment by querolita** [comment]
## Kimchi

### Configuration

The constraint system [`Builder`](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/constraints.rs#L197) contains a field called `runtime_tables` which is an optional vector of `RuntimeTableCfg` (one per table). Each of these [configurations](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/runtime_tables.rs#L25) are structs composed by a table identifier and a vector of field elements representing the indices of the desired table. The length of that vector defines the maximum length of the table. That is illustrated in the struct [`RuntimeTableSpec`](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/runtime_tables.rs#L14).
```rust 
pub struct RuntimeTableCfg<F> {
    pub id: i32,
    pub first_column: Vec<F>,
}
``` 
In order to set runtime tables one needs to create the constraint system like so: 
```rust
        let cs = ConstraintSystem::<G::ScalarField>::create(gates)
            .lookup(lookup_tables)
            .runtime(runtime_table_cfgs)
            .public(public)
            .prev_challenges(prev_challenges)
            .disable_gates_checks(disable_gates_checks)
            .max_poly_size(override_srs_size)
            .build()
            .unwrap();
```
Inside [`build`](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/constraints.rs#L883), the function [`create`](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/index.rs#L203) in `LookupConstraintSystem`(https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/index.rs#L160) will set the specification of the configured tables. In particular, after checking there are no duplicate IDs and the length is correct, it computes the runtime [selectors](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/index.rs#L285) for the rows where there are runtime tables (using for this an offset going from the lengths of the fixed tables and the length of the runtime's). After that's done, an actual [lookup table is created](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/index.rs#L323) with all entries set to zeros, and pushed into the list of lookup tables used. 

### Instantiation

Assigning actual data to each of the indices in the `first_column` of the `RuntimeTableCfg` is done by creating a [`RuntimeTable`](https://github.com/o1-labs/proof-systems/blob/0cd54de466460a51741fcc0f455a8cce38b6cbac/kimchi/src/circuits/lookup/runtime_tables.rs#L61) struct containing the table identifier and the second column containing the data. 
```rust
pub struct RuntimeTable<F> {
    pub id: i32,
    pub data: Vec<F>,
}
```
Once the table is created, the prover needs to load it. This is done by passing them as an input of the `ProverProof::create_recursive` function

### Querying

## `ocaml-types`

Some of these structs have an Ocaml counterpart.

**Comment by mitschabaude** [comment]
Note that there are bindings for adding a runtime table config, see Snarky.gates

**Comment by querolita** [comment]
@mitschabaude Indeed! I was digging into the actual runtime table and how to pass it to pickles though, since the config is not enough for it to work...
